components/core/src/DictionaryWriter.hpp

```diff
@@ -151,60 +141,59 @@ void DictionaryWriter<DictionaryIdType, EntryType>::close () {
     m_dictionary_compressor.close();
     m_dictionary_file_writer.close();
 
-    // Delete entries and clear maps
-    for (const auto& value_entry_pair : m_value_to_entry) {
-        delete value_entry_pair.second;
-    }
-    m_id_to_entry.clear();
-    m_value_to_entry.clear();
+    m_value_to_id.clear();
 
     m_is_open = false;
 }
 
 template <typename DictionaryIdType, typename EntryType>
-const EntryType* DictionaryWriter<DictionaryIdType, EntryType>::get_entry (DictionaryIdType id) const {
-    if (false == m_is_open) {
-        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
-    }
-
-    if (m_id_to_entry.count(id) == 0) {
-        throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
-    }
-    return m_id_to_entry.at(id);
-}
-
-template <typename DictionaryIdType, typename EntryType>
-void DictionaryWriter<DictionaryIdType, EntryType>::write_uncommitted_entries_to_disk () {
+void DictionaryWriter<DictionaryIdType, EntryType>::write_header_and_flush_to_disk () {
     if (false == m_is_open) {
         throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
     }
 
-    if (m_uncommitted_entries.empty()) {
-        // Nothing to do
-        return;
-    }
-
-    for (auto entry : m_uncommitted_entries) {
-        entry->write_to_file(m_dictionary_compressor);
-    }
-
     // Update header
     auto dictionary_file_writer_pos = m_dictionary_file_writer.get_pos();
     m_dictionary_file_writer.seek_from_begin(0);
-    m_dictionary_file_writer.write_numeric_value<uint64_t>(m_id_to_entry.size());
+    m_dictionary_file_writer.write_numeric_value<uint64_t>(m_value_to_id.size());
     m_dictionary_file_writer.seek_from_begin(dictionary_file_writer_pos);
 
     m_segment_index_compressor.flush();
     m_segment_index_file_writer.flush();
     m_dictionary_compressor.flush();
     m_dictionary_file_writer.flush();
-    m_uncommitted_entries.clear();
+}
+
+template <typename DictionaryIdType, typename EntryType>
+void DictionaryWriter<DictionaryIdType, EntryType>::preload_dictionary_value_to_map (streaming_compression::zstd::Decompressor& dictionary_decompressor,
+                                                                                     FileReader& dictionary_file_reader) {
+
+    auto num_dictionary_entries = read_dictionary_header(dictionary_file_reader);
+    EntryType entry;
+    for (size_t i = 0; i < num_dictionary_entries; ++i) {
+        entry.read_from_file(dictionary_decompressor);
+        const auto& str_value = entry.get_value();
+        if (m_value_to_id.find(str_value) != m_value_to_id.end()) {
+            SPDLOG_ERROR("entry value already exists");
+            throw OperationFailed(ErrorCode_Corrupt, __FILENAME__, __LINE__);
+        }
+        if (m_next_id > m_max_id) {
+            SPDLOG_ERROR("DictionaryWriter ran out of IDs.");
+            throw OperationFailed(ErrorCode_OutOfBounds, __FILENAME__, __LINE__);
+        }
+
+        DictionaryIdType id = m_next_id;
```

Comment:
what I understand is that with code below, you are taking care of the case where the id in the preloaded dictionary might be non-sequential (otherwise it's safe to assign m_next_id = num_dictionary_entries).
`if (id >= m_next_id) {
    m_next_id = id + 1;
}`

However, in decompression, we seem to assume that the entries are read in with sequentially increasing ID. so I would prefer to use this assumption in the preload as well.

Reply:
Hm, you're right. Past Kirk evidently forgot to add a TODO, sorry.

Okay, in that case, we could write `m_next_id = m_value_to_id.size()` after the loop rather than updating it on every iteration of the loop.